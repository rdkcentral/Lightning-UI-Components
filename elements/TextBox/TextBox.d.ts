import type { __TextType } from '@lightningjs/core';
import Base from '../../Base';
import { WithThemeStylesConstructor } from '../../mixins/withThemeStyles';

export interface TextBoxStyles {
  maxLinesSuffix: string;
  offsetY: number;
  offsetX: number;
  textStyle: __TextType;
  typography: Record<string, __TextType>;
}

declare const TextBox_base: WithThemeStylesConstructor<
  typeof Base,
  TextBoxStyles
>;

export default class TextBox extends TextBox_base {
  content?: string;
  maxLines?: number;
  maxLinesSuffix?: __TextType['maxLinesSuffix'];
  textStyle?: __TextType | string;
  textColor?: __TextType['textColor'];
  textAlign?: __TextType['textAlign'];
  verticalAlign?: __TextType['verticalAlign'];
  wordWrap?: __TextType['wordWrap'];
  wordWrapWidth?: __TextType['wordWrapWidth'];
}
