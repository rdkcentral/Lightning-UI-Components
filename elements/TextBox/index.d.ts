import lng from '@lightningjs/core';
import Base from '../../Base';
import { WithThemeStylesConstructor } from '../../mixins/withThemeStyles';
import { TextContent } from '../../layout/InlineContent';

export interface TextBoxStyles {
  maxLinesSuffix: string;
  offsetY: number;
  offsetX: number;
  textStyle: lng.textures.TextTexture.Settings;
  typography: Record<string, lng.textures.TextTexture.Settings>;
}

declare const TextBox_base: WithThemeStylesConstructor<
  typeof Base,
  TextBoxStyles
>;

export default class TextBox extends TextBox_base {
  content?: string | TextContent[];
  maxLines?: number;
  maxLinesSuffix?: lng.textures.TextTexture.Settings['maxLinesSuffix'];
  textStyle?: lng.textures.TextTexture.Settings | string;
  textColor?: lng.textures.TextTexture.Settings['textColor'];
  textAlign?: lng.textures.TextTexture.Settings['textAlign'];
  verticalAlign?: lng.textures.TextTexture.Settings['verticalAlign'];
  wordWrap?: lng.textures.TextTexture.Settings['wordWrap'];
  wordWrapWidth?: lng.textures.TextTexture.Settings['wordWrapWidth'];
}
