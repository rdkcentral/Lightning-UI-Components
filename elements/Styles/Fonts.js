/**
 * Copyright 2020 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * Fonts
 *
 * Contains global font style information to easily maintain consistency throughout components.
 */
export const TYPOGRAPHY = {
  display1: {
    fontSize: 56,
    lineHeight: 72,
    fontWeight: 'bold',
    letterSpacing: -0.4,
    verticalAlign: 'middle'
  },
  display2: {
    fontSize: 48,
    lineHeight: 64,
    fontWeight: 'bold',
    letterSpacing: -0.4,
    verticalAlign: 'middle'
  },
  headline1: {
    fontSize: 36,
    lineHeight: 48,
    fontWeight: 'bold',
    letterSpacing: -0.2,
    verticalAlign: 'middle'
  },
  headline2: {
    fontSize: 32,
    lineHeight: 40,
    fontWeight: 'bold',
    letterSpacing: -0.2,
    verticalAlign: 'middle'
  },
  headline3: {
    fontSize: 28,
    lineHeight: 36,
    fontWeight: 'bold',
    letterSpacing: -0.2,
    verticalAlign: 'middle'
  },
  body1: {
    fontSize: 32,
    lineHeight: 40,
    fontWeight: 500,
    letterSpacing: 0,
    verticalAlign: 'middle'
  },
  body2: {
    fontSize: 28,
    lineHeight: 36,
    fontWeight: 500,
    letterSpacing: 0,
    verticalAlign: 'middle'
  },
  body3: {
    fontSize: 24,
    lineHeight: 32,
    fontWeight: 500,
    letterSpacing: 0,
    verticalAlign: 'middle'
  },
  button1: {
    fontSize: 28,
    lineHeight: 36,
    fontWeight: 'bold',
    letterSpacing: -0.2,
    verticalAlign: 'middle'
  },
  button2: {
    fontSize: 24,
    lineHeight: 32,
    fontWeight: 'bold',
    letterSpacing: -0.2,
    verticalAlign: 'middle'
  },
  callout1: {
    fontSize: 24,
    lineHeight: 32,
    fontWeight: 'bold',
    letterSpacing: 2,
    textTransform: 'uppercase',
    verticalAlign: 'middle'
  },
  caption1: {
    fontSize: 20,
    lineHeight: 24,
    fontWeight: 'bold',
    letterSpacing: 0,
    verticalAlign: 'middle'
  }
};
