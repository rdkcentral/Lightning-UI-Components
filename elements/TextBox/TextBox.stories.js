import lng from '@lightningjs/core';
import TextBox from '.';
import mdx from './TextBox.mdx';
import context from '../../context';
import lightningbolt from '../../assets/images/ic_lightning_white_32.png';
import { getHexColor } from '../../utils';
import inlineContentStory from '../../layout/InlineContent/InlineContent.stories';

export default {
  title: 'Elements / TextBox',
  parameters: {
    docs: {
      page: mdx
    }
  }
};

const { args: inlineContentArgs, argTypes: inlineContentArgTypes } =
  inlineContentStory;

const lorum =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum sodales est eu eleifend interdum. Vivamus egestas maximus elementum. Sed condimentum ligula justo, non sollicitudin lectus rutrum vel. Integer iaculis vitae nisl quis tincidunt. Sed quis dui vehicula, vehicula felis a, tempor leo. Fusce tincidunt, ante eget pretium efficitur, libero elit volutpat quam, sit amet porta tortor odio non ligula. Ut sed dolor eleifend massa auctor porttitor eget ut lectus. Vivamus elementum lorem mauris, eu luctus tortor posuere sit amet. Nunc a interdum metus.';

export const Base = () =>
  class Base extends lng.Component {
    static _template() {
      return {
        TextBox: {
          type: TextBox
        }
      };
    }
  };

Base.argTypes = {
  content: {
    control: 'text',
    defaultValue: lorum,
    description: 'Text to be displayed in element',
    type: 'string',
    table: {
      defaultValue: { summary: '' }
    }
  },
  textAlign: {
    control: {
      type: 'select',
      options: ['left', 'center', 'right']
    },
    defaultValue: 'left',
    description: 'the horizontal alignment of TextBox content',
    type: 'string',
    table: {
      defaultValue: { summary: 'left' }
    }
  },
  textColor: {
    control: 'color',
    defaultValue: '#ffffff',
    description: 'Desired color of rendered text.',
    type: 'string',
    table: {
      defaultValue: { summary: '#ffffff' }
    }
  },
  textStyle: {
    control: {
      type: 'select',
      options: Object.keys(context.theme.typography)
    },
    defaultValue: 'body1',
    description:
      'Style of text to be displayed. This value will take precedent over the `textStyle` style property. Avalilable values display1, display2, headline1, headline2, headline3, body1, body2, body3, button1, button2, callout1, caption1. Alternatively an object can be passed in containing any [properties that the Lightning text texture supports](https://lightningjs.io/docs/#/lightning-core-reference/RenderEngine/Textures/Text?id=properties).',
    type: 'string',
    table: {
      defaultValue: { summary: 'body1' }
    }
  },
  maxLines: {
    control: {
      type: 'number',
      min: 0,
      max: 100
    },
    defaultValue: 3,
    description: 'Maximum number of lines to display before truncation',
    type: 'number',
    table: {
      defaultValue: { summary: 0 }
    }
  },
  maxLinesSuffix: {
    control: {
      type: 'text'
    },
    defaultValue: '...',
    description:
      'string to be displayed at the end of the line when truncation is used',
    type: 'string',
    table: {
      defaultValue: { summary: '...' }
    }
  },
  verticalAlign: {
    control: {
      type: 'select',
      options: ['top', 'middle', 'bottom']
    },
    defaultValue: 'middle',
    description: 'The vertical alignment of TextBox content.',
    type: 'string',
    table: {
      defaultValue: { summary: 'middle' }
    }
  },
  wordWrap: {
    control: 'boolean',
    defaultValue: true,
    description:
      'Allows long sentences to be broken into new lines when the width of TextBox content exceeds the wordWrapWidth',
    type: 'boolean',
    table: {
      defaultValue: { summary: true }
    }
  },
  wordWrapWidth: {
    control: {
      type: 'number',
      min: 0,
      max: 2000
    },
    defaultValue: 600,
    description: 'Width set to start wrapping to next line',
    type: 'number',
    table: {
      defaultValue: { summary: 0 }
    }
  }
};

export const WithInlineContentArray = () =>
  class WithInlineContentArray extends lng.Component {
    static _template() {
      return {
        TextBox: {
          type: TextBox,
          x: 200,
          w: 400,
          content: [
            'Text',
            {
              icon: lightningbolt,
              color: getHexColor('00ff00'),
              title: 'Green Lightning Bolt'
            },
            'Wrapping Text',
            {
              icon: 'http://myriad.merlin.comcast.com/select/logo?entityId=8527084350383982239&width=32&height=&ratio=1x1&trim=false',
              title: 'Rotten Tomatoes rating'
            },
            'and more text',
            {
              text: 'with some red ',
              style: { textColor: getHexColor('FF6194') }
            },
            'for fun',
            { badge: 'HD', title: 'HD' },
            { badge: 'SD', title: 'SD' }
          ]
        }
      };
    }
  };

WithInlineContentArray.args = inlineContentArgs;
WithInlineContentArray.argTypes = inlineContentArgTypes;

export const WithInlineContentString = () =>
  class WithInlineContentArray extends lng.Component {
    static _template() {
      return {
        TextBox: {
          type: TextBox,
          w: 400,
          content:
            'This is an example of using custom markup {ICON:settings|http://myriad.merlin.comcast.com/select/logo?entityId=8527084350383982239&width=32&height=&ratio=1x1&trim=false} with a linebreak{NEWLINE}{BADGE:HD} that includes all available types like {TEXT:styled text|italic}.',
          textStyles: {
            italic: {
              fontStyle: 'italic',
              textColor: getHexColor('FF6194')
            }
          }
        }
      };
    }
  };

WithInlineContentString.args = inlineContentArgs;
WithInlineContentString.argTypes = inlineContentArgTypes;
