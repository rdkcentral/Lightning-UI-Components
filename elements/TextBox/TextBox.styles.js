export default theme => {
  return {
    typography: theme.typography,
    maxLinesSuffix: '...'
  };
};
