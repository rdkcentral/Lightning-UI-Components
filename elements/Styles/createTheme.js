/**
 * Copyright 2020 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import lng from '@lightningjs/core';
import { clone } from '../../utils';
import { CORNER_RADIUS } from './Styles';
import { COLORS_NEUTRAL, getHexColor } from './Colors';

function glow({
  w,
  h,
  color = 0xffffffff,
  borderRadius = CORNER_RADIUS.small,
  blur = spacing(2)
}) {
  return {
    color: color,
    mount: 0.5,
    x: w / 2,
    y: h / 2 + blur / 2,
    zIndex: -1,
    texture: lng.Tools.getShadowRect(
      w - spacing(2),
      h - spacing(2),
      borderRadius,
      blur
    )
  };
}

function luminance({ w, h, blur = 3, padding = spacing(8), texture = null }) {
  return {
    type: lng.components.FastBlurComponent,
    x: w / 2,
    y: h / 2 + spacing(3),
    mount: 0.5,
    w,
    h,
    amount: blur,
    padding: padding,
    content: {
      Image: {
        w,
        h,
        texture: texture
      }
    },
    zIndex: -1
  };
}

const materials = {
  glow,
  luminance
};

function spacing(multiplier) {
  return 8 * multiplier;
}

const gradientColor = COLORS_NEUTRAL.light2;
const gradientAnimation = {
  duration: 0.6,
  actions: [
    {
      p: 'colorUl',
      v: {
        0: getHexColor(gradientColor, 72),
        1: getHexColor(gradientColor, 56)
      }
    },
    {
      p: 'colorUr',
      v: {
        0: getHexColor(gradientColor, 24),
        1: getHexColor(gradientColor, 16)
      }
    },
    {
      p: 'colorBr',
      v: { 0: 0x00, 1: getHexColor(gradientColor, 0) }
    },
    {
      p: 'colorBl',
      v: {
        0: getHexColor(gradientColor, 24),
        1: getHexColor(gradientColor, 16)
      }
    }
  ]
};

const SIZES = {
  icon: {
    small: 32,
    medium: 40,
    large: 48
  }
};

/**
 * Combines two theme objects to create a new theme
 * @param {Object|function} theme - An object or callback that accepts the default theme as an argument. Gets merged with the default theme
 * @param {Object} [defaultTheme] - Theme to be merged with.
 */
export default (theme, defaultTheme = {}) => {
  if (typeof theme === 'function') {
    return clone(defaultTheme, theme(defaultTheme));
  }

  return clone(defaultTheme, theme);
};
