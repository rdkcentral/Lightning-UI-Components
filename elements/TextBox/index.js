import Base from '../../Base';
import styles from './TextBox.styles';
import withStyles from '../../mixins/withThemeStyles';
import { getValidColor } from '../../Styles';
import context from '../../context';
export default class TextBox extends withStyles(Base, styles) {
  static _template() {
    return {
      alpha: 0.001,
      text: { textBaseline: 'bottom' }
    };
  }

  static get __componentName() {
    return 'TextBox';
  }

  static get tags() {
    return ['TextBox'];
  }

  static get properties() {
    return [
      'content',
      'textStyle',
      'textColor',
      'textAlign',
      'verticalAlign',
      'wordWrap',
      'maxLines',
      'wordWrapWidth',
      'maxLinesSuffix'
    ];
  }

  _construct() {
    super._construct && super._construct();
    context.on('fontsLoaded', this._forceUpdate.bind(this));
  }

  _detach() {
    context.off('fontsLoaded', this._forceUpdate.bind(this));
  }

  _init() {
    this.on('txLoaded', this._setDimensions.bind(this));
  }

  _forceUpdate() {
    // Fonts that are loaded asynchronous via the theme are not picked up by the component unless the textureSourceHashmap is cleared first.
    const id = this.text._getLookupId();
    if (this.text && id) {
      this.stage.textureManager.textureSourceHashmap.delete(id);
      this.text._changed();
    }
  }

  _setDimensions() {
    const width = this.texture.getRenderWidth();
    const height = this.texture.getRenderHeight();
    if (width && height) {
      this.h = height;
      this.w = width;
      // Position updates can produce flash of poorly positioned content, hide the element until measurements are made.
      if (1 > this.alpha) {
        this.patch({
          alpha: 1
        });
      }

      this._notifyAncestors();
    }
  }

  _setContent(content) {
    if ('string' !== typeof content) {
      return '';
    }
    return content;
  }

  _setTextColor(textColor) {
    const color = getValidColor(textColor);
    if (null === color) {
      return undefined;
    }
    return color;
  }

  _setTextAlign(textAlign) {
    const options = ['left', 'center', 'right'];
    if ('string' !== typeof textAlign || -1 === options.indexOf(textAlign)) {
      return undefined;
    }
    return textAlign;
  }

  _setVerticalAlign(verticalAlign) {
    const options = ['top', 'middle', 'bottom'];
    if (
      'string' !== typeof verticalAlign ||
      -1 === options.indexOf(verticalAlign)
    ) {
      return undefined;
    }
    return verticalAlign;
  }

  _setMaxLines(maxLines) {
    if ('number' !== typeof maxLines || 0 > maxLines) {
      return undefined;
    }
    return maxLines;
  }

  _setWordWrap(wordWrap) {
    if ('boolean' !== typeof wordWrap) {
      return undefined;
    }
    return wordWrap;
  }

  _setWordWrapWidth(wordWrapWidth) {
    if ('number' !== typeof wordWrapWidth || 0 > wordWrapWidth) {
      return undefined;
    }
    return wordWrapWidth;
  }

  _setMaxLinesSuffix(maxLinesSuffix) {
    if ('string' !== typeof maxLinesSuffix) {
      return this._componentStyles.maxLinesSuffix;
    }
    return maxLinesSuffix;
  }

  get title() {
    return this._content;
  }

  _notifyAncestors() {
    this.fireAncestors('$itemChanged');
    this.signal('textBoxChanged', {
      w: this.w,
      h: this.h
    });
  }

  _update() {
    const fontStyle = {
      ...(this._componentStyles.typography[this.textStyle] ||
        this._componentStyles.typography.body1)
    };

    this.constructor.properties.forEach(prop => {
      if ('fontStyle' !== prop && 'undefined' !== typeof this[`_${prop}`]) {
        const key = 'content' === prop ? 'text' : prop;

        fontStyle[key] = this[`_${prop}`];
      }
    });

    this.patch({ text: fontStyle });
  }

  get announce() {
    return this._announce || this.content;
  }

  set announce(announce) {
    super.announce = announce;
  }
}
