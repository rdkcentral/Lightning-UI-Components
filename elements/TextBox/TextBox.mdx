import { Canvas, Story } from '@storybook/addon-docs';
import { Theme } from '../../.storybook/utils';
import Tile from '.';

# TextBox

General purpose TextBox component

## Usage

Set a content property with your text and specify a property style as a string. Typeography styles can be found in @lightning/ui/Styles/Fonts.js

```js
import { TextBox } from '@lightning/ui/elements';

TextBox: {
  type: TextBox,
  content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
  style: 'headline1',
  w: 600,
}
```

<Canvas>
  <Story id="elements-textbox--base" />
</Canvas>

### Properties

| name           | type          | required | default                | description                                                                                                                                                                                                                                                                                                                                                                                                                                         |
| -------------- | ------------- | -------- | ---------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| content        | string        | true     | -                      | Text to be displayed in element                                                                                                                                                                                                                                                                                                                                                                                                                     |
| maxLines       | number        | false    | 0                      | Maximum number of lines to display before truncation.                                                                                                                                                                                                                                                                                                                                                                                               |
| maxLinesSuffix | string        | false    | '...'                  | String to be displayed at the end of the line when truncation is used                                                                                                                                                                                                                                                                                                                                                                               |
| textAlign      | string        | false    | left                   | Avalilable values 'left', 'right', or 'center'.                                                                                                                                                                                                                                                                                                                                                                                                     |
| textColor      | hex           | false    | #ffffff                | Desired text color.                                                                                                                                                                                                                                                                                                                                                                                                                                 |
| textStyle      | string/object | false    | theme.typography.body1 | Style of text to be displayed. This value will take precedent over the `textStyle` style property. Avalilable values display1, display2, headline1, headline2, headline3, body1, body2, body3, button1, button2, callout1, caption. Alternatively an object can be passed in containing any [properties that the Lightning text texture supports](https://lightningjs.io/docs/#/lightning-core-reference/RenderEngine/Textures/Text?id=properties). |
| verticalAlign  | string        | false    | middle                 | When lineHeight is larger than fontSize, specifies how text should align vertically inside each line. Avalilable values 'top', 'middle', or 'bottom'.                                                                                                                                                                                                                                                                                               |
| wordWrap       | boolean       | false    | true                   | Allows long sentences to be broken into new lines.                                                                                                                                                                                                                                                                                                                                                                                                  |
| wordWrapWidth  | number        | false    | 0                      | Width set to start wrapping to next line.                                                                                                                                                                                                                                                                                                                                                                                                           |

TextBox also supports rendering as an InlineContent component. See [InlineContent documentation](/?path=/docs/layout-inlinecontent--basic) for table of supported properties.

### Style Properties

| name           | type   | description                                                                                                                                                                                                                                             |
| -------------- | ------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| maxLinesSuffix | string | String to be displayed at the end of the line when truncation is used.                                                                                                                                                                                  |
| offsetX        | number | Used to ensure text is positioned correctly                                                                                                                                                                                                             |
| offsetY        | number | Used to ensure text is positioned correctly                                                                                                                                                                                                             |
| textStyle      | object | Default style of text to be displayed. Object should come from the current theme or contain any properties that the [Lightning text texture supports](https://lightningjs.io/docs/#/lightning-core-reference/RenderEngine/Textures/Text?id=properties). |
| typography     | object | The typography object set on the current theme. Pass the key as string to `textStyle` to use the associated text style on the typography object.                                                                                                        |
