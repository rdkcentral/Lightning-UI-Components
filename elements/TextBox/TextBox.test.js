import TestUtils from '../../test/lightning-test-utils';
import TextBox from '.';
import context from '../../context';
import { TYPOGRAPHY } from '../../Styles';
import { getValidColor } from '../../utils';
import XfinityTheme from '../../themes/xfinity';

const createElement = TestUtils.makeCreateComponent(TextBox);

const testOptions = async (element, optionProp, optionsValues, match) => {
  for (const option of optionsValues) {
    await (() => {
      return new Promise(resolve => {
        element.content = 'Hello world';
        element[optionProp] = option;
        setTimeout(() => {
          let expects;
          switch (typeof match) {
            case 'function':
              expects = match(option);
              break;
            case 'undefined':
              expects = option;
              break;
            default:
              expects = match;
          }
          expect(element.tag('Text').text[optionProp]).toBe(expects);
          resolve();
        });
      });
    })();
  }
  return;
};

describe('TextBox', () => {
  let element, testRenderer;

  beforeEach(async () => {
    context.setTheme(XfinityTheme);
    [element, testRenderer] = createElement({}, { spyOnMethods: ['_update'] });
    await element.__updateSpyPromise;
  });

  afterEach(() => {
    element = null;
    testRenderer = null;
  });

  describe('basic', () => {
    it('renders', () => {
      const tree = testRenderer.toJSON(2);
      expect(tree).toMatchSnapshot();
    });

    it('renders with content prop', async () => {
      element.content = 'Hello world';
      await element.__updateSpyPromise;
      expect(element._Text.text.text).toBe('Hello world');
    });

    it('should set Text and InlineContent to undefined if the content prop is invalid', () => {
      element.content = null;
      testRenderer.forceAllUpdates();
      expect(element._Text).toBeUndefined();
      expect(element._InlineContent).toBeUndefined();

      expect(element.w).toBe(0);
    });

    it('should set width and height to 0 if the content prop is invalid', () => {
      element.content = null;
      testRenderer.forceAllUpdates();

      expect(element.w).toBe(0);
      expect(element.h).toBe(0);
    });

    it('renders with a title that matches the content prop', async () => {
      element.content = 'Hello world';
      await element.__updateSpyPromise;
      expect(element.title).toBe('Hello world');
    });

    it('should log a warning if the user attempts to smooth any values to the TextBox', async () => {
      const warnMock = jest.fn();
      context.warn = warnMock;
      element.smooth = { w: 100 };
      expect(warnMock.mock.calls.length).toBe(1);
    });
  });

  describe('styling', () => {
    it('should fallback to "body1" object if the style prop is an invalid string', async () => {
      element.content = 'Hello world';
      element.textStyle = 'invalidstyle';
      await element.__updateSpyPromise;
      expect(element._Text).toEqual(
        expect.objectContaining({
          text: expect.objectContaining(XfinityTheme.typography.body1)
        })
      );
    });

    it('should set style to "body1" object if the style prop is null', async () => {
      element.content = 'Hello World';
      element.textStyle = null;
      await element.__updateSpyPromise;
      expect(element._Text).toEqual(
        expect.objectContaining({
          text: expect.objectContaining(XfinityTheme.typography.body1)
        })
      );
    });

    it('should set style to "body1" object if the style prop is not a string or object', async () => {
      element.content = 'Hello World';
      element.textStyle = () => {};
      await element.__updateSpyPromise;
      expect(element._Text).toEqual(
        expect.objectContaining({
          text: expect.objectContaining(XfinityTheme.typography.body1)
        })
      );
    });

    it('should allow the default textStyle to be set with componentStyle in the theme', async () => {
      context.updateTheme({
        componentStyle: {
          TextBox: {
            defaultTextStyle: 'headline1'
          }
        }
      });
      await element.__updateSpyPromise;
      element.content = 'Hello world';
      await element.__updateSpyPromise;
      expect(element._Text.text.fontSize).toBe(36);
    });

    it('should accept a hex color for the textColor prop', async () => {
      const validColors = [
        '#F44336',
        '#FFEBEE',
        '#FFCDD2',
        '#EF9A9A',
        '#E57373',
        '#EF5350',
        '#F44336',
        '#E53935',
        '#D32F2F',
        '#C62828'
      ];
      await testOptions(element, 'textColor', validColors, getValidColor);
    });

    it('should set textColor to white if the prop is invalid', async () => {
      const invalidColors = [
        'fff',
        true,
        false,
        () => {},
        undefined,
        null,
        'red'
      ];
      await testOptions(
        element,
        'textColor',
        invalidColors,
        getValidColor('#ffffff')
      );
    });
  });

  describe('alignment and formatting', () => {
    it('should accept left, right, & center as options for textAlign', async () => {
      [element, testRenderer] = createElement({
        style: Object.keys(TYPOGRAPHY)[0]
      });
      await testOptions(element, 'textAlign', ['right', 'center', 'left']);
    });

    it('should fallback to default for textAlign if options are not supported in the text texture', async () => {
      await testOptions(
        element,
        'textAlign',
        ['justify', 100, -100, 0, 1, true, false],
        'left'
      );
    });

    it('should accept top, middle, & bottom as options for verticalAlign', async () => {
      await testOptions(element, 'verticalAlign', ['bottom', 'middle', 'top']);
    });

    it('should fallback to the default for verticalAlign if options are not supported in the text texture', async () => {
      const { verticalAlign: defaultVerticalAlign } = TYPOGRAPHY.body1;
      await testOptions(
        element,
        'verticalAlign',
        ['center', true, false, 0, 1, () => {}],
        defaultVerticalAlign
      );
    });

    it('should accept boolean value for wordWrap option', async () => {
      await testOptions(element, 'wordWrap', [false, true]);
    });

    it('should set default wordWrap value to true if boolean not passed as option', async () => {
      await testOptions(
        element,
        'wordWrap',
        [0, 1, 'true', () => {}, null, undefined],
        true
      );
    });

    it('should accept maxLines option', async () => {
      await testOptions(
        element,
        'maxLines',
        new Array(10).fill().map((v, i) => i + 1)
      );
    });

    it('should only accept maxLines option values that are positive numbers where n > 0', async () => {
      await testOptions(
        element,
        'maxLines',
        [-1, true, false, null, undefined, () => {}],
        0
      );
    });

    it('should accept wordWrapWidth option', async () => {
      await testOptions(
        element,
        'wordWrapWidth',
        Array(10)
          .fill()
          .map((v, i) => i + 100)
      );
    });

    it('should only accept wordWrapWidth option values that are positive numbers where n > 0', async () => {
      await testOptions(
        element,
        'wordWrapWidth',
        [-1, true, false, null, undefined, () => {}],
        0
      );
    });

    it('should accept maxLinesSuffix option of a string', async () => {
      await testOptions(element, 'maxLinesSuffix', ['...', '>>>', '???']);
    });

    it('should fallback to default for maxLinesSuffix if prop is anything other than a string', async () => {
      await testOptions(
        element,
        'maxLinesSuffix',
        [-1, true, false, null, undefined, () => {}],
        '...'
      );
    });
  });

  describe('inline content', () => {
    it('should parse an array of content and render as InlineContent', async () => {
      const content = [
        'Text',
        {
          icon: 'http://myriad.merlin.comcast.com/select/logo?entityId=8527084350383982239&width=32&height=&ratio=1x1&trim=false',
          title: 'Rotten Tomatoes rating'
        },
        { badge: 'HD', title: 'HD' },
        { badge: 'SD', title: 'SD' }
      ];
      [element, testRenderer] = createElement(
        { content },
        { spyOnMethods: ['_update'] }
      );
      await element.__updateSpyPromise;
      expect(element._isInlineContent).toBe(true);
      expect(element._InlineContent.content).toBe(content);
    });

    it('should parse a string with markup and render as InlineContent', async () => {
      const content =
        'Example {ICON:settings|http://myriad.merlin.comcast.com/select/logo?entityId=8527084350383982239&width=32&height=&ratio=1x1&trim=false} with a linebreak{NEWLINE}{BADGE:HD} that includes {TEXT:styled text|italic}.';
      [element, testRenderer] = createElement(
        { content },
        { spyOnMethods: ['_update'] }
      );
      await element.__updateSpyPromise;
      expect(element._isInlineContent).toBe(true);
      expect(element._InlineContent.content).toBe(content);
    });

    it('should announce its content', async () => {
      const content = 'Hello world';
      [element, testRenderer] = createElement({ content });
      expect(element.announce).toBe(content);
    });
  });

  describe('marquee', () => {
    it('should create a Marquee element when marquee is true and content longer than wordWrapWidth', async () => {
      element.content = 'hello world';
      element.wordWrapWidth = 20;
      element.marquee = true;
      await element.__updateSpyPromise;
      expect(element._Marquee).toBeDefined();

      [element, testRenderer] = createElement(
        {
          content: [
            'Text',
            {
              icon: 'http://myriad.merlin.comcast.com/select/logo?entityId=8527084350383982239&width=32&height=&ratio=1x1&trim=false',
              title: 'Rotten Tomatoes rating'
            },
            { badge: 'HD', title: 'HD' },
            { badge: 'SD', title: 'SD' }
          ],
          wordWrapWidth: 20,
          marquee: true
        },
        { spyOnMethods: ['_update'] }
      );
      await element.__updateSpyPromise;
      expect(element._Marquee).toBeDefined();
    });

    it('should toggle alpha values of Marquee and Text components based on the marquee property', async () => {
      element.content = 'hello world';
      element.wordWrapWidth = 20;
      element.marquee = false;
      await element.__updateSpyPromise;
      expect(element._Marquee).toBe(undefined);
      expect(element._Text.alpha).toBe(1);

      element.marquee = true;
      await element.__updateSpyPromise;
      expect(element._Marquee.alpha).toBe(1);
      expect(element._Text.alpha).toBe(0.001);

      element.marquee = false;
      await element.__updateSpyPromise;
      expect(element._Marquee.alpha).toBe(0.001);
      expect(element._Text.alpha).toBe(1);
    });

    it('marquee component should only ever have maxLines of 1', async () => {
      element.content = 'hello world';
      element.maxLines = 3;
      element.wordWrapWidth = 20;
      element.marquee = true;
      await element.__updateSpyPromise;
      expect(element._Marquee.title.maxLines).toBe(1);
    });
  });
});
