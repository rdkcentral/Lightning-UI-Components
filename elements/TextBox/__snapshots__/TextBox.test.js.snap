// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextBox renders 1`] = `
Object {
  "Component": Object {
    "active": false,
    "alpha": 0.001,
    "attached": true,
    "boundsMargin": null,
    "clipping": false,
    "color": 4294967295,
    "enabled": true,
    "flex": false,
    "flexItem": false,
    "h": 0,
    "hasFinalFocus": true,
    "hasFocus": true,
    "isComponent": true,
    "mount": 0,
    "mountX": 0,
    "mountY": 0,
    "pivot": 0.5,
    "pivotX": 0.5,
    "pivotY": 0.5,
    "ref": "Component",
    "renderOfScreen": undefined,
    "renderToTexture": false,
    "scale": 1,
    "scaleX": 1,
    "scaleY": 1,
    "state": "",
    "tag": [Function],
    "tags": Array [
      "Component",
    ],
    "texture": Object {
      "fontFace": "DMSansRegular",
      "fontSize": 32,
      "lineHeight": 40,
      "textBaseline": "bottom",
      "type": "TextTexture",
      "verticalAlign": "middle",
    },
    "type": "TextBox",
    "visible": true,
    "w": 0,
    "x": 0,
    "y": 0,
    "zIndex": 0,
  },
}
`;
